src/module.c

```diff
@@ -6038,6 +6023,103 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
     return reply;
 }
 
+/* Exported API to call any Redis command from modules
+ *
+ * * **rm_user**: a RedisModuleUser object the call should be performed as for ACL enforcement purposes
+ * * **cmdname**: The Redis command to call.
+ * * **fmt**: A format specifier string for the command's arguments. Each
+ *   of the arguments should be specified by a valid type specification. The
+ *   format specifier can also contain the modifiers `!`, `A`, `3` and `R` which
+ *   don't have a corresponding argument.
```

Comment:
i'd like to see how we can arrange the code in a way that won't cause git to think all these lines were deleted and re-added.
it's hard to review, and may cause unnecessary conflicts.

Reply:
ok, so that would probably mean, renaming RM_Call to the internal name, so only  change that line, and then create a new RM_Call elsewhere?

Reply:
maybe.. you'll probably need some forward declarations.
maybe there's a way to break the code into functions and lay the lines of the code as close as how they were.
keep in mind that the comment needs to be right next to RM_Call in order to make it in the generated docs.
anyway, it doesn't have to be perfect, but let's at least try to get something better.

Reply:
ok, based on discussion below, removed / revert all the WithUser changes and now have a RM_SetContextModuleUser that if set will cause RM_Call to run the client as that.

The issue we have is how this will work if the underlying user is freed on you somewhere else (i.e. users aren't referenced counted, maybe they should be?).  A thought to solve this, would be that the expectation is that you do a 

RM_SetContextModuleUser() and then an immediate RM_Call() each time and the user is then invalidated after used by the RM_Call, so if you did

RM_SetContextModuleUser()
RM_Call()
RM_Call()

this second call would run as the root/null user.  therefore don't really have lifetime management issues.

Reply:
i don't think we should revert it after RM_Call, i think we should revert it when the context is destroyed.
maybe on long lived contexts we have an issue, but then don't we have that issue anyway?

p.s. shouldn't it be RM_SetModuleContextUser (i.e. ModuleContext instead of ContextModule)?

Reply:
going backwards:

I was thinking of it as Set the Context's ModuleUser (as it's called a RedisModuleUser struct).

In general we don't have "free floating" user objects like this.  i.e. it be attached to a client (and then it invalidate/disconnect the client when we fee the object).  In this case, I dont think its safe to keep it valid long term.

With that said, the same argument can be made I guess if this done in an external thread, then any use of RedisModuleUser is probably just as problematic as can be freed on main thread without one's knowledge in thread.